src/anemoi/training/losses/mse.py

```diff
@@ -89,3 +89,109 @@ def forward(
+        data_variances : Optional[torch.Tensor], optional
+            precomputed, per-variable stepwise variance estimate, by default None
+        ignore_nans : bool, optional
+            Allow nans in the loss and apply methods ignoring nans for measuring the loss, by default False
+        """
+        super().__init__()
+
+        self.avg_function = torch.nanmean if ignore_nans else torch.mean
+        self.sum_function = torch.nansum if ignore_nans else torch.sum
+
+        self.inside_lam = inside_lam
+        self.wmse_contribution = wmse_contribution
+        self.register_buffer("weights", node_weights, persistent=True)
+        self.register_buffer("weights_inside_lam", node_weights[mask], persistent=True)
+        self.register_buffer("weights_outside_lam", node_weights[~mask], persistent=True)
+        self.register_buffer("mask", mask, persistent=True)
+        if data_variances is not None:
+            self.register_buffer("ivar", data_variances, persistent=True)
+
+    def forward(
+        self,
+        pred: torch.Tensor,
+        target: torch.Tensor,
+        squash: bool = True,
+    ) -> torch.Tensor:
+        """Calculates the lat-weighted MSE loss.
+
+        Parameters
+        ----------
+        pred : torch.Tensor
+            Prediction tensor, shape (bs, lat*lon, n_outputs)
+        target : torch.Tensor
+            Target tensor, shape (bs, lat*lon, n_outputs)
+        squash : bool, optional
+            Average last dimension, by default True
+
+        Returns
+        -------
+        torch.Tensor
+            Weighted MSE loss
+        """
+        full_out_dims = pred[:, :, :, 0]
+
+        if self.inside_lam:
+            pred = pred[:, :, self.mask]
+            target = target[:, :, self.mask]
+            weights_selected = self.weights_inside_lam
+        else:
+            pred = pred[:, :, ~self.mask]
+            target = target[:, :, ~self.mask]
+            weights_selected = self.weights_outside_lam
+
+        out = torch.square(pred - target)
+
+        # Use variances if available
+        if hasattr(self, "ivar"):
+            out *= self.ivar
+
+        # Squash by last dimension
+        if squash:
+            out = self.avg_function(out, dim=-1)
+            # Weight by area
+            out = out * weights_selected.expand_as(out)
+            if self.wmse_contribution:
+                out /= self.sum_function(self.weights.expand_as(full_out_dims))
+            else:
+                out /= self.sum_function(weights_selected.expand_as(out))
+            return self.sum_function(out)
+
+        # Weight by area
+        out = out * weights_selected[..., None].expand_as(out)
+        if self.wmse_contribution:
+            out /= self.sum_function(self.weights[..., None].expand_as(full_out_dims))
+        else:
+            out /= self.sum_function(weights_selected[..., None].expand_as(out))
+        return self.sum_function(out, axis=(0, 1, 2))
```

Comment:
If the `mask` is to be used within the `scalars`, this will need to be refactored. I am happy to propose a refactor if you can provide some expected results to test against.

Reply:
Some changes are being made to the scalar that will assist in the refactor of this code. https://github.com/ecmwf/anemoi-training/pull/137